File: Vigenere/Rust/vigenere_test.rs
#[cfg(test)]
mod tests {
    use super::vigenere::Vigenere;

    fn make_table() -> [[char; 26]; 26] {
        [['a'; 26]; 26]
    }

    #[test]
    fn encode_simple() {
        let mut table = make_table();
        let v = Vigenere::new("attackatdawn".to_string(), "lemon".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "lxfopvefrnhr");
    }

    #[test]
    fn encode_mixed_case() {
        let mut table = make_table();
        let v = Vigenere::new("AttackAtDawn".to_string(), "LEMON".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "lxfopvefrnhr");
    }

    #[test]
    fn encode_with_spaces_and_symbols() {
        let mut table = make_table();
        let v = Vigenere::new("attack at dawn!".to_string(), "lemon".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "lxfopv efrnhr!");
    }

    #[test]
    fn encode_single_char_key() {
        let mut table = make_table();
        let v = Vigenere::new("abcxyz".to_string(), "b".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "bcdyza");
    }

    #[test]
    fn encode_long_key() {
        let mut table = make_table();
        let v = Vigenere::new("dog".to_string(), "superlongkey".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "vcl");
    }

    #[test]
    fn encode_empty_input() {
        let mut table = make_table();
        let v = Vigenere::new("".to_string(), "key".to_string(), &mut table);
        assert_eq!(v.encode(), "");
    }

    #[test]
    fn encode_uppercase_key() {
        let mut table = make_table();
        let v = Vigenere::new("hello".to_string(), "Z".to_string(), &mut table);
        let encoded = v.encode();
        assert_eq!(encoded, "gdkkn");
    }
}

File: Vigenere/C++/Vigenere_Test.cpp
#include <gtest/gtest.h>
#include "Vigenere.h"

TEST(VigenereTest, EncodeDecodeBasic) 
{
    vigenere v("attackatdawn", "lemon");
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "lxfopvefrnhr");
    EXPECT_EQ(v.decode(encoded), "attackatdawn");
}

TEST(VigenereTest, EncodeWithUppercase) 
{
    vigenere v("AttackAtDawn", "LEMON");
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "LxfopvEfRnhr");
    EXPECT_EQ(v.decode(encoded), "AttackAtDawn");
}

TEST(VigenereTest, EncodeWithSpacesAndPunctuation) 
{
    vigenere v("hello, world!", "key");
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "rijvs, uyvjn!");
    EXPECT_EQ(v.decode(encoded), "hello, world!");
}

TEST(VigenereTest, LongKeyShortText) 
{
    vigenere v("hi", "superlongkey");
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "zj");
    EXPECT_EQ(v.decode(encoded), "hi");
}

TEST(VigenereTest, SingleLetterKey) 
{
    vigenere v("abcxyz", "b"); 
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "bcdyza");
    EXPECT_EQ(v.decode(encoded), "abcxyz");
}

TEST(VigenereTest, EmptyInput) 
{
    vigenere v("", "key");
    EXPECT_EQ(v.encode(), "");
    EXPECT_EQ(v.decode(""), "");
}

TEST(VigenereTest, MixedCaseAndSymbols) 
{
    vigenere v("TeSt123!", "key");
    std::string encoded = v.encode();
    EXPECT_EQ(encoded, "DiCx123!");
    EXPECT_EQ(v.decode(encoded), "TeSt123!");
}
